### PR TITLE
[KYUUBI #5677][AUTHZ] Typeof expression miss column information

### LINKED ISSUE #5677: [Bug] Typeof expression miss column information
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the bug

Typeof expression miss column information

### Affects Version(s)

master, 1.8.0

### Kyuubi Server Log Output

_No response_

### Kyuubi Engine Log Output

_No response_

### Kyuubi Server Configurations

_No response_

### Kyuubi Engine Configurations

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [X] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to fix.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### _Why are the changes needed?_
To close #5677 
Typeof expression miss column information, in this pr we add a place holder to support this.

For sql 
```
SELECT typeof(id), typeof(age) FROM default.t1 LIMIT 1
```

Without this pr, typeof after optimizer passed to PrivilegeBuild is just an attribute, miss the origin child expression info.
```
GlobalLimit 21
+- LocalLimit 21
   +- Project [int AS typeof(id)#27, int AS typeof(age)#28]
      +- Relation default.t1[id#18,age#19,part#20] parquet
```

When handle this plan's project list, it's an attribute and miss expression when expression result is a constant value, then we can't extract the `TypeOf`'s child expression.
image

This is caused by `TypeOf`  expression is foldable
image

It will be convert to a constant value after spark optimizer, then cause we miss the child expression, then can't extract columns by `collectLeaves`

In this pr we wrap the `TypeOf` by a non-foldable expression placeholder then we can get the expression contains column of table when `mergeProjection`
After this pr, the plan is like below
```
GlobalLimit 21
+- LocalLimit 21
   +- Project [typeofplaceholder(id#21) AS typeof(id)#30, typeofplaceholder(day#23) AS typeof(day)#31]
      +- HiveTableRelation [`default`.`table1`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, Data Cols: [id#21, scope#22, day#23], Partition Cols: []]

```

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### _Was this patch authored or co-authored using generative AI tooling?_
No

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @yaooqinn

**Comment by yaooqinn** [comment]
What is the problem here?

**Comment by AngersZhuuuu** [comment]
> What is the problem here?

Without this pr, it will check the privilege of all table, not column

**Comment by yaooqinn** [comment]
I see the id, age field in the project. Why not pull them out?

**Comment by AngersZhuuuu** [comment]
> I see the id, age field in the project. Why not pull them out?

Yea

**Comment by AngersZhuuuu** [comment]
Without this pr the UT's result is  as below cc @yaooqinn 
```
"Permission denied: user [someone] does not have [select] privilege on [default/table1]" did not contain "does not have [select] privilege on [default/table1/id,default/table1/day]"
ScalaTestFailureLocation: org.apache.kyuubi.plugin.spark.authz.ranger.HiveCatalogRangerSparkExtensionSuite at (RangerSparkExtensionSuite.scala:1316)
org.scalatest.exceptions.TestFailedException: "Permission denied: user [someone] does not have [select] privilege on [default/table1]" did not contain "does not have [select] privilege on [default/table1/id,default/table1/day]"
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)

```

**Comment by yaooqinn** [comment]
> > I see the id, age field in the project. Why not pull them out?
> 
> Yea 截屏2023-11-13 下午2 01 10


Why not pull them out?

**Comment by AngersZhuuuu** [comment]
> > > I see the id, age field in the project. Why not pull them out?
> > 
> > 
> > Yea 截屏2023-11-13 下午2 01 10
> 
> Why not pull them out?

The name and expr id is not same

**Comment by yaooqinn** [comment]
Why is typeof special, or how many are they?

**Comment by AngersZhuuuu** [comment]
> Why is typeof special, or how many are they?

Since typeof's result only needs the expression's datatype, it doesn't need real value.  
Only see one such case:
1. need expression
2. Result is constant

**Comment by yaooqinn** [comment]
> Since typeof's result only needs the expression's datatype, it doesn't need real value.


Shall we bypass it?

**Comment by AngersZhuuuu** [comment]
> > Since typeof's result only needs the expression's datatype, it doesn't need real value.
> 
> Shall we bypass it?

But it need table's read privilege, can't just ignore, better extract the correct column?

**Comment by yaooqinn** [comment]
It's much more like `explain select a from src` to me

**Comment by AngersZhuuuu** [comment]
> It's much more like `explain select a from src` to me

But only when the query's result is empty, then this sql have the output. If there is no row, show an empty result

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala [hunk 1]
```diff
@@ -1296,4 +1296,37 @@ class HiveCatalogRangerSparkExtensionSuite extends RangerSparkExtensionSuite {
...
+          doAs(
+            someone,
+            sql(
+              s"""
+                 |SELECT
+                 |typeof(cast(id as string)),
+                 |typeof(substring(day, 1, 3))
+                 |FROM $db1.$table1""".stripMargin).show()))(
+          s"does not have [select] privilege on [$db1/$table1/id,$db1/$table1/day]")
+        doAs(admin, sql(s"SELECT typeof(id), typeof(day) FROM $db1.$table1").show())
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala:1328]: check values

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala [hunk 2]
```diff
@@ -1296,4 +1296,37 @@ class HiveCatalogRangerSparkExtensionSuite extends RangerSparkExtensionSuite {
...
+          sql(
+            s"""
+               |CREATE TABLE IF NOT EXISTS $db1.$table1(
+               |id int,
+               |scope int,
+               |day string)
+               |""".stripMargin))
+        doAs(admin, sql(s"INSERT INTO $db1.$table1 SELECT 1, 2, 'TONY'"))
+        interceptContains[AccessControlException](
+          doAs(someone, sql(s"SELECT typeof(id), typeof(day) FROM $db1.$table1").show()))(
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala:1316]: do not use show

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala (hunk 1 above)
Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala:1328]: SELECT typeof(typeof)

typeof typeof(id + age)

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala]: Done

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala]: How about current?

**Comment by bowenliang123** [comment]
Could you add the example for the plan in either the issue or the pr description to show the problem and the plan?

**Comment by AngersZhuuuu** [comment]
> Could you add the example for the plan in either the issue or the pr description to show the problem and the plan?

How about current?

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/rule/RuleEliminateTypeOf.scala [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.plugin.spark.authz.rule
+
+import org.apache.spark.sql.catalyst.expressions.TypeOf
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+
+import org.apache.kyuubi.plugin.spark.authz.rule.typeof.TypeOfPlaceHolder
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/rule/RuleEliminateTypeOf.scala:24]: typeof -> expression

Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/rule/RuleEliminateTypeOf.scala]: > typeof -> expression

Done

**Comment by AngersZhuuuu** [comment]
Any more suggestion?

**Review by yaooqinn — APPROVED** [review]

**Comment by bowenliang123** [comment]
> > Could you add the example for the plan in either the issue or the pr description to show the problem and the plan?
> 
> How about current?

Sorry, it's still too difficult for me. I'm out of the context.
Show the SQL or situation first, then the plan, and the suggested solution.

**Comment by bowenliang123** [comment]
If you don't have time next time, there's no need to ping me. Lately, I haven't been keeping up with your series of PRs. They always get merged quickly, and my suggestions are being skipped. Not enough information for me to understand.  If the Kyuubi community is okay with this 'no-explanation-just-go' approach, I don't have any objections. I attribute all of my willingness to my own shortcomings in abilities.

**Comment by AngersZhuuuu** [comment]
> > > Could you add the example for the plan in either the issue or the pr description to show the problem and the plan?
> > 
> > 
> > How about current?
> 
> Sorry, it's still too difficult for me. I'm out of the context. Show the SQL or situation first, then the plan, and the suggested solution.

Add explain about root cause and solution.

**Pull request closed without merging**